### PR TITLE
Exposing DynamicMethod.DefineParameter API and adding test

### LINKED ISSUE #29715: Why is DynamicMethod.DefineParameter not exposed?
I was trying to run [the `DynamicMethod` example on docs.MS](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.dynamicmethod?view=netcore-2.0#examples) on .Net Core 2.x, but it failed with:

> Program.cs(60,15): error CS1061: 'DynamicMethod' does not contain a definition for 'DefineParameter' and no extension method 'DefineParameter' accepting a first argument of type 'DynamicMethod' could be found (are you missing a using directive or an assembly reference?)
>
> Program.cs(61,15): error CS1061: 'DynamicMethod' does not contain a definition for 'DefineParameter' and no extension method 'DefineParameter' accepting a first argument of type 'DynamicMethod' could be found (are you missing a using directive or an assembly reference?)

This is consistent with API reference, [which says that `DynamicMethod.DefineParameter` does not exist on .Net Core](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.dynamicmethod.defineparameter?view=netframework-4.7.2#applies-to).

On the other hand, [the method does exist in the source code](https://github.com/dotnet/coreclr/blob/ba110d5/src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.cs#L522).

Is there a specific reason why the method is not exposed?

This might be a duplicate of https://github.com/dotnet/corefx/issues/18072.

### PR DESCRIPTION
**Pull request by maryamariyan**
`DynamicMethod.DefineParameter` was already implemented but not exposed.
This PR:
- Exposes `DefineParameter`
- Adds test to exercise the code. It asserts the parameter name, attribute and type

Fixes #29715 

cc: @AtsushiKan @jkotas @danmosemsft

### CODE REVIEW

File: src/System.Reflection.Emit.Lightweight/tests/DynamicMethodDefineParameter.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            ILGenerator il = powerOf.GetILGenerator(256);
+            il.Emit(OpCodes.Ldarg_0);
+            il.Emit(OpCodes.Ldarg_1);
+            il.Emit(OpCodes.Call, typeof(Math).GetMethod("Pow"));
+            il.Emit(OpCodes.Ret);
+
+            var paramNames = new string[] { "param1", "param2" };
+            for (int i = 0; i < paramNames.Length; i++)
+            {
+                powerOf.DefineParameter(i+1, ParameterAttributes.In, paramNames[i]);
```
Comment (danmosemsft) [inline src/System.Reflection.Emit.Lightweight/tests/DynamicMethodDefineParameter.cs]: Nit, spacing around `+`

**Comment by danmosemsft** [comment]
@maryamariyan if you need more tasks msybe you could take https://github.com/dotnet/corefx/issues/12055 then.

**Comment by snakefoot** [comment]
@maryamariyan Is it possible to get an updated nuget-package released with DefineParameter included:

https://www.nuget.org/packages/System.Reflection.Emit.Lightweight/

It would really help at my company to start adoption of NetCore, by adding NetStandard2 support for our basic libraries.

**Comment by snakefoot** [comment]
@maryamariyan + @karelz  + @danmosemsft Is it possible to get an updated nuget-package released with DefineParameter included:

https://www.nuget.org/packages/System.Reflection.Emit.Lightweight/

It would really help at my company to start adoption of NetCore, by adding NetStandard2 support for our basic libraries.

**Comment by karelz** [comment]
It seems we are just fixing the ref assembly. The API is present in .NET Core 2.0:
https://github.com/dotnet/coreclr/blob/release/2.0.0/src/mscorlib/src/System/Reflection/Emit/DynamicMethod.cs#L529

And it is available on .NET Framework 2.0+: https://apisof.net/catalog/System.Reflection.Emit.DynamicMethod.DefineParameter(Int32,ParameterAttributes,String)

Not sure how costly it is for us to ship packages off cycle though. Moreover from master branch (the package is not built/published in .NET Core 2.0 or 2.1 branches).

@snakefoot did you consider workarounds to light up via Reflection? The method is public and will work on platforms which have `DynamicMethod`.

**Comment by maryamariyan** [comment]
you can get the package from myget here: 
https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Reflection.Emit.Lightweight

If you wanted to use it for netcore, you would need to use .netcoreapp30 and use a preview version of it

You can follow instructions from source-build repo [here](https://github.com/dotnet/source-build/tree/master/Documentation) and use NETCoreAppMaximumVersion 

example gist from @safern is here:
https://gist.github.com/maryamariyan/ba630243a30bed874afe9dab59d4551d

**Comment by snakefoot** [comment]
@maryamariyan Thank you for the preview package, will try it out. Any ETA for when there will be an official release on nuget?

**Comment by karelz** [comment]
.NET Core 3.0 is expected to ship in Q1 2019: https://github.com/dotnet/core/blob/master/roadmap.md#upcoming-ship-dates (the dates can change of course).

**Comment by snakefoot** [comment]
@karelz No chance this low-risk feature can be promoted to an earlier release? Is it impossible to use the nuget-package without the application is a NetCore3-application?

**Comment by safern** [comment]
> Is it impossible to use the final nuget-package without the application is a NetCore3-application?

I just doubled check the PR and the description of the PR and the implementation for this API is available since .netcore2.0 -- however since we never exposed this API in the reference assembly until now the only way to use it is to reference the preview package from myget that @maryamariyan pointed to. However the library/application that references this package needs to target either .netstandard2.0 or .netcoreapp3.0.

If you use it from a .netstandard2.0 library and then that library is used from .netcore2.0 or later then you will be fine. You can test it with the myget package to confirm

However if you want to reference this package directly from a .netcore application and use this API you will not be able to build successfully unless you target netcoreapp3.0.

> No chance this low-risk feature can be promoted to an earlier release?

Unfortunately, that will not be possible, since this is not a bug fix and it is exposing a new API this has to wait until we fork the master branch to our new release branch, and that will happen next until 3.0

**Comment by snakefoot** [comment]
> If you use it from a .netstandard2.0 library and then that library is used from .netcore2.0 or later then you will be fine. You can test it with the myget package to confirm

That is fine with me. When can I expect a final release of this nuget-package? Since it is not coupled to NetCore3 release.

**Comment by safern** [comment]
> When can I expect a final release of this nuget-package? Since it is not coupled to NetCore3 release.

It is indeed coupled with NetCore3 release since the reference assembly is part of the net core shared framework (is part of the Microsoft.NETCore.App) package -- and we can't ship a final package from master, it has to be from one of our servicing branches (i.e release/2.1 branch), which only accept bug fixes, but not API additions.

**Comment by snakefoot** [comment]
> It is indeed coupled with NetCore3 release since the reference assembly is part of the net core shared framework (is part of the Microsoft.NETCore.App) package

Not sure I understand. You just said the following:

> If you use it from a .netstandard2.0 library and then that library is used from .netcore2.0 or later then you will be fine.

And how you say that it only works from NetCore3 ?

**Comment by safern** [comment]
> And how you say that the nuget-package only works from NetCore3-Apps ?

If your application's target framework is netcoreapp2.1 and you try to reference `DynamicMethod.DefineParameter` you will not be able to build even if you reference the package. The reason is because the reference assembly (which is the contract of the APIs available for that library) is part of the Microsoft.NETCore.App package. This API wasn't exposed in the reference assembly of .netcoreapp2.0 nor .netcoreapp2.1. So in order to use this API from a .netcore application directly you would have to do it from a .netcoreapp3.0 application.

The only way to be able to use it from a 2.0/2.1 application is indirectly, with a bridge library that targets .netstandard2.0

```cs
// this method is inside a .netstandard2.0 library
public void UseApi()
{
         ...
         DynamicMethod.DefineParameter usage here.
}
```

then from a 2.0/2.1 application you could reference this library and call `UseApi` method.

That is why this is indirectly coupled to netcore3 -- eventhough, if it wasn't coupled, it would be the same outcome, since we never release a final nuget package without an official release just because it is low-risk. We have our release guidelines and we have to follow them accordingly.

**Comment by snakefoot** [comment]
So I can build a NetStandard2.0 library that depends on this package. And if this NetStandard2.0 package is then used by a NetCore2-App then the application will at some point in runtime explode into flames because DefineParameter does not exist?

**Comment by safern** [comment]
No, that is actually what is supported.

What is not supported is build a netcoreapp2.0/2.1 application that depends on this preview-nuget-package and use DefineParameter, you will not be able to build the application.

**Comment by snakefoot** [comment]
So my NetStandard2.0 library will also not compile when using the preview-nuget-package? Will there come a NetStandard3.0 together with NetCore3.0 ?

If so should this preview-nuget-package not include NetStandard3.0 as targetframework?

**Comment by safern** [comment]
> So my NetStandard2.0 library will also not compile when using the preview-nuget-package?

This case actually would work. Here is an example on how you can use it: https://github.com/safern/DefineParametersDemo 

So the supported scenarios with this preview nuget package are the following:

- Using it to build a .NET Standard 2.0 library that references `DefineParameter` api -- then consume that library from any .NET Core application
- Using it to build a .NET Core 3.0 application that references `DefineParameter` api. 

Not supported:
- Try to use `DefineParameter` api directly from a .NET Core 2.0/2.1 application.

> Will there come a NetStandard3.0 together with NetCore3.0 ?

No, there are no plans of shipping a .NET Standard 3.0 from what I know.

**Comment by snakefoot** [comment]
> Using it to build a .NET Standard 2.0 library that references DefineParameter api -- then consume that library from any .NET Core application

Think we are walking in circles here. But lets do this one more time. I just want to use this nuget-package in my own netstandard2.0-library. And you say this is possible and my own netstandard2.0-library will work just fine in any NetCoreApp2.0. Why can I not get a official release of this nuget-package then?

**Comment by safern** [comment]
> Why can I not get a official release of this nuget-package then?

You will be able to get an official release of this nuget package whenever we release .NET Core 3.0.

We release all the official nuget packages with our .NET Core official releases. We don't do standalone package releases. So we are not able to just do a release for an individual package, that is why you will have to wait until .NET Core 3.0 release. I'm sorry about that ☹️

**Comment by snakefoot** [comment]
So the nuget-package has no dependency on the NetCore3-Api, but for some reason ver. 4.6.0 of the package has a dependency on NetCore3-release-date.

Not possible to release a ver. 4.5.0 of this nuget-package with this method included?

**Comment by safern** [comment]
> Not possible to release a ver. 4.5.0 of this nuget-package with DefineParameter-method included?

Nope, the only way to release a 4.5.0 version of this package would need to be from the release/2.1.0 branch https://github.com/dotnet/corefx/tree/release/2.1.0 -- and in that branch we didn't build that package yet. This package was added like 2 months ago into master branch. 

If you look at the project in release/2.1.0 branch it doesn't have a pkg folder: https://github.com/dotnet/corefx/tree/release/2.0.0/src/System.Reflection.Emit.Lightweight

If we wanted to release a 4.5.0 version of this package we would need to port this PR and the addition of the pkg folder to the release/2.1.0 branch, which is not acceptable by our release guidelines, since a release/x branches are closed, and we only accept bug fixes there, no API additions or new packages.

**Comment by snakefoot** [comment]
@safern Well thank you for the explanation. Would have been easier if you have just said that nuget-package comes from the same branch as the NetCore3, and it is not possible to backport to an earlier branch unless bugfix (Didn't know that the source code for all NetCore-packages resides in the same repository-branch)

Well just have to sit and wait then :)

**Pull request merged**